libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,469 @@
+	//       not only for builtins.
+	if (auto const* funCall = get_if<FunctionCall>(&_exprStmt.expression))
+		if (BuiltinFunction const* builtin = m_dialect.builtin(funCall->functionName.name))
+			if (builtin->controlFlowSideEffects.terminates)
+			{
+				m_currentBlock->exit = CFG::BasicBlock::Terminated{};
+				m_currentBlock = &m_graph.makeBlock();
+			}
+}
+
+void ControlFlowGraphBuilder::operator()(Block const& _block)
+{
+	ScopedSaveAndRestore saveScope(m_scope, m_info.scopes.at(&_block).get());
+	for (auto const& statement: _block.statements)
+		std::visit(*this, statement);
+}
+
+void ControlFlowGraphBuilder::operator()(If const& _if)
+{
+	auto&& [ifBranch, afterIf] = makeConditionalJump(std::visit(*this, *_if.condition));
+	m_currentBlock = ifBranch;
+	(*this)(_if.body);
+	jump(*afterIf);
+}
+
+void ControlFlowGraphBuilder::operator()(Switch const& _switch)
+{
+	yulAssert(m_currentBlock, "");
+	auto ghostVariableId = m_graph.ghostVariables.size();
+	YulString ghostVariableName("GHOST[" + to_string(ghostVariableId) + "]");
+	auto& ghostVar = m_graph.ghostVariables.emplace_back(Scope::Variable{""_yulstring, ghostVariableName});
+
+	// Artificially generate:
+	// let <ghostVariable> := <switchExpression>
+	VariableSlot ghostVarSlot{ghostVar, debugDataOf(*_switch.expression)};
+	m_currentBlock->operations.emplace_back(CFG::Operation{
+		Stack{std::visit(*this, *_switch.expression)},
+		Stack{ghostVarSlot},
+		CFG::Assignment{_switch.debugData, {ghostVarSlot}}
+	});
+
+	BuiltinFunction const* equalityBuiltin = m_dialect.equalityFunction({});
+	yulAssert(equalityBuiltin, "");
+
+	// Artificially generate:
+	// eq(<literal>, <ghostVariable>)
+	auto makeValueCompare = [&](Literal const& _value) {
+		yul::FunctionCall const& ghostCall = m_graph.ghostCalls.emplace_back(yul::FunctionCall{
+			_value.debugData,
+			yul::Identifier{{}, "eq"_yulstring},
+			{_value, Identifier{{}, ghostVariableName}}
+		});
+		CFG::Operation& operation = m_currentBlock->operations.emplace_back(CFG::Operation{
+			Stack{ghostVarSlot, LiteralSlot{valueOfLiteral(_value), _value.debugData}},
+			Stack{TemporarySlot{ghostCall, 0}},
+			CFG::BuiltinCall{_switch.debugData, *equalityBuiltin, ghostCall, 2},
+		});
+		return operation.output.front();
+	};
+	CFG::BasicBlock& afterSwitch = m_graph.makeBlock();
+	yulAssert(!_switch.cases.empty(), "");
+	for (auto const& switchCase: _switch.cases | ranges::views::drop_last(1))
+	{
+		yulAssert(switchCase.value, "");
+		auto&& [caseBranch, elseBranch] = makeConditionalJump(makeValueCompare(*switchCase.value));
+		m_currentBlock = caseBranch;
+		(*this)(switchCase.body);
+		jump(afterSwitch);
+		m_currentBlock = elseBranch;
+	}
+	Case const& switchCase = _switch.cases.back();
+	if (switchCase.value)
+	{
+		CFG::BasicBlock& caseBranch = m_graph.makeBlock();
+		makeConditionalJump(makeValueCompare(*switchCase.value), caseBranch, afterSwitch);
+		m_currentBlock = &caseBranch;
+	}
```

Comment:
For 

```
{
    sstore(0, 0)
    switch sload(0)
    case 0 {
        sstore(0x01, 0x0101)
    }
    sstore(0x04, 0x0101)
}
```

In the above example, the nonZero case jumps to the `sstore(0x04, 0x0101)` block, whereas the `zero` 'directly' goes to the `store(0x04, 0x0101)` block. Shouldn't it be the other way around (based on how current codegen does it)?

I'm also okay with always having jumps in all cases, and deciding how to lay out code so that we can skip one of the jumps during the code transform.

See [visualizer](https://dreampuf.github.io/GraphvizOnline/#%20%20%20digraph%20CFG%20%7B%0A%20%20%20%20nodesep%3D0.7%3B%0A%20%20%20%20node%5Bshape%3Dbox%5D%3B%0A%0A%20%20%20%20Entry%20%5Blabel%3D%22Entry%22%5D%3B%0A%20%20%20%20Entry%20-%3E%20Block0%3B%0A%20%20%20%20Block0%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x00%200x00%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20sload%3A%20%5B%200x00%20%5D%20%3D%3E%20%5B%20TMP%5Bsload%2C%200%5D%20%5D%5Cl%5C%0A%20%20%20%20Assignment(GHOST%5B0%5D)%3A%20%5B%20TMP%5Bsload%2C%200%5D%20%5D%20%3D%3E%20%5B%20GHOST%5B0%5D%20%5D%5Cl%5C%0A%20%20%20%20eq%3A%20%5B%20GHOST%5B0%5D%200x00%20%5D%20%3D%3E%20%5B%20TMP%5Beq%2C%200%5D%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block0%20-%3E%20Block0Exit%3B%0A%20%20%20%20Block0Exit%20%5Blabel%3D%22%7B%20TMP%5Beq%2C%200%5D%7C%20%7B%20%3C0%3E%20Zero%20%7C%20%3C1%3E%20NonZero%20%7D%7D%22%20shape%3DMrecord%5D%3B%0A%20%20%20%20Block0Exit%3A0%20-%3E%20Block1%3B%0A%20%20%20%20Block0Exit%3A1%20-%3E%20Block2%3B%0A%0A%20%20%20%20Block1%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x0101%200x04%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block1Exit%20%5Blabel%3D%22MainExit%22%5D%3B%0A%20%20%20%20Block1%20-%3E%20Block1Exit%3B%0A%0A%20%20%20%20Block2%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x0101%200x01%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block2%20-%3E%20Block2Exit%20%5Barrowhead%3Dnone%5D%3B%0A%20%20%20%20Block2Exit%20%5Blabel%3D%22Jump%22%20shape%3Doval%5D%3B%0A%20%20%20%20Block2Exit%20-%3E%20Block1%3B%0A%0A%20%20%20%20%7D%0A)

Reply:
The yul optimizer should transform that switch to an ``if`` anyways, so I'd tend to say better to not introduce special cases in the graph or in the code transform - at least for now, if it turns out things like that would better be done here or during code generation, we can still do it later on - there's generally loads of optimization opportunities left in the entire process.

Reply:
That being said, the code generator even now won't actually generate that jump :-).
The logic there is to just keep generating through jumps, unless the block to be jumped to is already generated (so the non-zero branch here will be staged for later generation and the zero branch will immediately follow the pre-switch branch without jump).

Reply:
Although it may actually even be possible to avoid the logic for that in the code generation step, generate that jump and hope that the legacy codegen inliner will remove the jump, but I haven't checked that that works in all cases :-).

Reply:
> That being said, the code generator even now won't actually generate that jump

```
======= <stdin> (EVM) =======

Pretty printed source:
object "object" {
    code {
        sstore(0, 0)
        switch sload(0)
        case 0 { sstore(5, 5) }
        sstore(0x04, 0x0101)
    }
}


Binary representation:
6000600055600054600081146012576018565b60056005555b50610101600455

Text representation:
    /* "<stdin>":16:17   */
  0x00
    /* "<stdin>":13:14   */
  0x00
    /* "<stdin>":6:18   */
  sstore
    /* "<stdin>":36:37   */
  0x00
    /* "<stdin>":30:38   */
  sload
    /* "<stdin>":48:49   */
  0x00
    /* "<stdin>":43:77   */
  dup2
  eq
  tag_2
  jumpi
    /* "<stdin>":23:77   */
  jump(tag_1)
    /* "<stdin>":43:77   */
tag_2:
    /* "<stdin>":69:70   */
  0x05
    /* "<stdin>":66:67   */
  0x05
    /* "<stdin>":59:71   */
  sstore
    /* "<stdin>":23:77   */
tag_1:
  pop
    /* "<stdin>":95:101   */
  0x0101
    /* "<stdin>":89:93   */
  0x04
    /* "<stdin>":82:102   */
  sstore
```

The old codegen doesn't jump in the nonZero case, whereas it jumps in the zero case. Currently the graph does the opposite, right?

Reply:
You'd get something like the following which is neither better nor worse:
```
    /* "<stdin>":16:17   */
  0x00
    /* "<stdin>":13:14   */
  0x00
    /* "<stdin>":6:18   */
  sstore
    /* "<stdin>":36:37   */
  0x00
    /* "<stdin>":30:38   */
  sload
    /* "<stdin>":48:49   */
  0x00
    /* "<stdin>":43:77   */
  dup2
  eq
  tag_2
  jumpi
    /* "<stdin>":23:77   */
tag_1:
  pop
    /* "<stdin>":95:101   */
  0x0101
    /* "<stdin>":89:93   */
  0x04
    /* "<stdin>":82:102   */
  sstore
  
  ...eventually "stop" ...
  
    /* "<stdin>":43:77   */
tag_2:
    /* "<stdin>":69:70   */
  0x05
    /* "<stdin>":66:67   */
  0x05
    /* "<stdin>":59:71   */
  sstore
    /* "<stdin>":23:77   */
  jump(tag_1)
```